### PR TITLE
feat: support any precision in PromQL

### PR DESCRIPTION
**Pull request by Taylor-lagrange**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

Support any precision in PromQL

Please explain IN DETAIL what the changes are in this PR and why they are needed:

When detect non-millisecond timestamp in PromQL planner on time index, cast it to ms.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [ ]  This PR does not require documentation updates.

### CODE REVIEW

File: src/promql/src/extension_plan/instant_manipulate.rs [hunk 1]
```diff
@@ -342,12 +342,14 @@ impl InstantManipulateStream {
...
-        // TODO(ruihang): maybe the input is not timestamp millisecond array
+
         let ts_column = input
             .column(self.time_index)
             .as_any()
             .downcast_ref::<TimestampMillisecondArray>()
-            .unwrap();
+            .ok_or(DataFusionError::Execution(
+                "Time index Column downcast to TimestampMillisecondArray failed".into(),
+            ))?;
```
Comment (evenyag) [inline src/promql/src/extension_plan/instant_manipulate.rs:352]:
```suggestion
            .ok_or_else(|| DataFusionError::Execution(
                "Time index Column downcast to TimestampMillisecondArray failed".into(),
            ))?;
```

File: src/promql/src/extension_plan/normalize.rs [hunk 2]
```diff
@@ -250,12 +250,13 @@ pub struct SeriesNormalizeStream {
...
     pub fn normalize(&self, input: RecordBatch) -> DataFusionResult<RecordBatch> {
-        // TODO(ruihang): maybe the input is not timestamp millisecond array
         let ts_column = input
             .column(self.time_index)
             .as_any()
             .downcast_ref::<TimestampMillisecondArray>()
-            .unwrap();
+            .ok_or(DataFusionError::Execution(
+                "Time index Column downcast to TimestampMillisecondArray failed".into(),
+            ))?;
```
Comment (evenyag) [inline src/promql/src/extension_plan/normalize.rs:259]:
```suggestion
            .ok_or_else(|| DataFusionError::Execution(
                "Time index Column downcast to TimestampMillisecondArray failed".into(),
            ))?;
```

File: src/promql/src/extension_plan/range_manipulate.rs [hunk 3]
```diff
@@ -472,12 +472,17 @@ impl RangeManipulateStream {
...
+        input: &RecordBatch,
+    ) -> DataFusionResult<(ArrayRef, Vec<(u32, u32)>)> {
         let ts_column = input
             .column(self.time_index)
             .as_any()
             .downcast_ref::<TimestampMillisecondArray>()
-            .unwrap();
+            .ok_or(DataFusionError::Execution(
+                "Time index Column downcast to TimestampMillisecondArray failed".into(),
+            ))?;
```
Comment (evenyag) [inline src/promql/src/extension_plan/range_manipulate.rs:485]:
```suggestion
            .ok_or_else(|| DataFusionError::Execution(
                "Time index Column downcast to TimestampMillisecondArray failed".into(),
            ))?;
```

File: src/promql/src/planner.rs [hunk 4]
```diff
@@ -910,9 +911,65 @@ impl PromPlanner {
...
+
+        let time_index_type = &table_schema
+            .timestamp_column()
+            .with_context(|| TimeIndexNotFoundSnafu {
+                table: table_ref.to_quoted_string(),
+            })?
+            .data_type;
+
+        let is_time_index_ms = *time_index_type == ConcreteDataType::time_millisecond_datatype()
+            || *time_index_type == ConcreteDataType::timestamp_millisecond_datatype();
```
Comment (evenyag) [inline src/promql/src/planner.rs:933]: Time index only allow the timestamp type.
```suggestion
        let is_time_index_ms = *time_index_type == ConcreteDataType::timestamp_millisecond_datatype();
```

File: src/promql/src/planner.rs [hunk 5]
```diff
@@ -910,9 +911,65 @@ impl PromPlanner {
...
+                .collect::<Vec<_>>();
+            scan_plan = LogicalPlanBuilder::from(scan_plan)
+                .project(expr)
+                .context(DataFusionPlanningSnafu)?
+                .build()
+                .context(DataFusionPlanningSnafu)?;
+        }
+
+        // Safety: `scan_filters` is not empty.
+        let result = LogicalPlanBuilder::from(scan_plan)
```
Comment (evenyag) [inline src/promql/src/planner.rs:972]: Could we add a test to validate the plan generated?

**Review by killme2008 — APPROVED** [review]
LGTM +:1

**Review by evenyag — APPROVED** [review]

**Pull request merged**